cpp/src/arrow/json/reader.cc

```diff
@@ -42,132 +42,443 @@ namespace arrow {
+  }
 
-    return MakeChunkedArrayBuilder(task_group_, pool_, promotion_graph, type, &builder_);
-  }
-
-  Status ParseAndInsert(const std::shared_ptr<Buffer>& partial,
-                        const std::shared_ptr<Buffer>& completion,
-                        const std::shared_ptr<Buffer>& whole, int64_t block_index) {
-    std::unique_ptr<BlockParser> parser;
-    RETURN_NOT_OK(BlockParser::Make(pool_, parse_options_, &parser));
-    RETURN_NOT_OK(parser->ReserveScalarStorage(partial->size() + completion->size() +
-                                               whole->size()));
-
-    if (partial->size() != 0 || completion->size() != 0) {
-      std::shared_ptr<Buffer> straddling;
-      if (partial->size() == 0) {
-        straddling = completion;
-      } else if (completion->size() == 0) {
-        straddling = partial;
-      } else {
-        ARROW_ASSIGN_OR_RAISE(straddling,
-                              ConcatenateBuffers({partial, completion}, pool_));
+  static Future<std::shared_ptr<StreamingReaderImpl>> MakeAsync(
+      std::shared_ptr<DecodeContext> context, std::shared_ptr<io::InputStream> stream,
+      io::IOContext io_context, Executor* cpu_executor, const ReadOptions& read_options) {
+    ARROW_ASSIGN_OR_RAISE(
+        auto buffer_it,
+        io::MakeInputStreamIterator(std::move(stream), read_options.block_size));
+    ARROW_ASSIGN_OR_RAISE(
+        auto buffer_gen,
+        MakeBackgroundGenerator(std::move(buffer_it), io_context.executor()));
+
+    AsyncGenerator<DecodedBlock> decoding_gen;
+    int max_readahead = 0;
+    if (read_options.use_threads) {
+      // Prepare a source generator capable of async-reentrancy and parallel execution
+      if (!cpu_executor) {
+        cpu_executor = GetCpuThreadPool();
       }
-      RETURN_NOT_OK(parser->Parse(straddling));
+      max_readahead = cpu_executor->GetCapacity();
+
+      // Since the chunking/decoding steps are heavy we want to schedule them as a
+      // separate task so as to maximize task distribution across CPU cores
+      //
+      // TODO: Add an `always_transfer` parameter to `MakeTransferredGenerator`?
+      buffer_gen = [source = std::move(buffer_gen), cpu_executor] {
+        return cpu_executor->TransferAlways(source());
+      };
+      auto chunking_gen = MakeChunkingGenerator(std::move(buffer_gen),
+                                                MakeChunker(context->parse_options()));
+
+      // Despite having already transferred to the CPU executor, we don't bother
+      // synchronizing access to the chunking generator because `MappingGenerator` queues
+      // jobs and keeps only one future from its source active at a time. This is also
+      // why we can apply readahead later despite the generator we're providing not being
+      // async-reentrant
+      //
+      // The subsequent decoding task should run on the same CPU thread as the chunking
+      // continuation. However, the next read can be initialized before then
```

Comment:
Does it mean that CPU-heavy work (chunking + decoding) is essentially single-threaded?

Reply:
It just means that a separate CPU thread won't be spawned to do the decoding - but due to the way `MappingGenerator` works, the decoding can still run in parallel. Each callback from its source starts the next pull before running the map function.

Reply:
In parallel with what, though? Decoding of different blocks should be done in multiple threads, while chunking is inherently serial.

Reply:
Only in parallel with other decoding tasks. The decoder just continues to use the same CPU thread that the chunking completed on while the next IO -> chunking job gets started up asynchronously. However, only one chunking job will be ongoing at a time despite the fact that it's on the CPU executor.

It's pretty difficult to explain without the implementation, but [here](https://github.com/benibus/arrow/blob/bc5ce43f01045f6d4e264329e3663c9b64cd6860/cpp/src/arrow/util/async_generator.h#L239) is where the next serial task gets triggered if there are still jobs in the queue. Note that this occurs in the callback of the prior chunking task, so we should already be on a CPU thread. Shortly after, [the decoding task would run in the same callback](https://github.com/benibus/arrow/blob/bc5ce43f01045f6d4e264329e3663c9b64cd6860/cpp/src/arrow/util/async_generator.h#L247). In the meantime, the next chunking task can be scheduled on a different CPU thread, which would then become _its_ decoding thread, and so on.

I suppose I could be wrong though, as this isn't exactly documented. I only really looked into it out of suspicion due to the way it was used in the CSV reader.

Reply:
Also, if what I said doesn't make much sense, here's a basic example (over a few iterations):
```
Reading: IoThread(?) --> Chunking: CpuThread(0) ... Decoding: CpuThread(0)
--------------------------------------------------------------------------
                                                    Decoding: CpuThread(0)
Reading: IoThread(?) --> Chunking: CpuThread(1) ... Decoding: CpuThread(1)
--------------------------------------------------------------------------
                                                    Decoding: CpuThread(0)
                                                    Decoding: CpuThread(1)
Reading: IoThread(?) --> Chunking: CpuThread(2) ... Decoding: CpuThread(2)
```
The next reading/chunking iteration starts _between_ the second and third phase, allowing the decoding tasks to gradually saturate multiple cores while still running the chunker serially.

Reply:
I see, thanks. This might be part of a comment somewhere to make things easier to understand.

Reply:
Yeah, I'll add something.

On an unrelated note, I'm pretty sure this is partially relevant to https://github.com/apache/arrow/issues/14792. The documentation for `MappingGenerator` heavily undersells its usefulness IMO.

Reply:
Thanks. Can you comment on #14792?

Reply:
Yeah, will do